Resolver middleware

stevelacey opened the pull request:
## Summary

GraphQL-aware middleware solve a variety of problems, a while ago I needed to be able to Logstash various stats about the queries going through my API (e.g. what args were passed, what fields were requested), and more recently I wanted to move authorization ahead of validation after I learnt about #481. Resolves #591.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

### Checklist
- [x] Existing tests have been adapted and/or new tests have been added
- [x] Add a CHANGELOG.md entry
- [x] Update the README.md
- [x] Code style has been fixed via `composer fix-style`

mfn commented:
Looks promising, can't wait for the tests.

I'm also interested how well this works together with "batched requests". Would be great to see this in tests.

Example scenario: a batch request with 2 graphql queries were the middleware decides on the first to skip it but the 2nd should still be resolved correctly.

mfn commented:
Small feedback: I'm all for consistency and formatting, but would be great if you could resist the urge to mix this in a PR introduces a new feature 😬

stevelacey commented:
@mfn let me know what you think, I'll add tests once we're happy with the implementation.

I figure I should probably add another way to define global middleware, I for one am using 2 base classes (one for query and one for mutation), and am sharing the `middleware` attribute definition via a trait, which is awkward, and doesn't seem to work once `Field` has a definition either.

Maybe `config`? Or `GraphQL::middleware([])`?

There's also some work to do here to distinguish the config of HTTP middleware.

stevelacey commented:
> Small feedback: I'm all for consistency and formatting, but would be great if you could resist the urge to mix this in a PR introducing a new feature 😬

Yah trying, every time I come to copy an existing example I find 20 variations 🙊

stevelacey commented:
@mfn separated out the README fixes https://github.com/rebing/graphql-laravel/pull/595

stevelacey commented:
@mfn added some tests for an `ExampleMiddleware` that does some things + the command

mfn commented:
@stevelacey much appreciated!

I'm a bit busy at the moment I would like to take my time with the review, also play around a bit because I might see also some use-cases I want to try out.

Definitely helps to already have tests in this stage to accelerate this 👍 

Please bear with me.

stevelacey commented:
@mfn no rush, please do take your time 👍

mfn commented on README.md:
````diff
@@ -662,6 +666,150 @@ class UsersQuery extends Query
+php artisan make:graphql:middleware ResolvePage
+```
+
+This command will place a new ResolvePage class within your app/Http/Middleware directory.
````
I think it's in a different dir/namespace
```suggestion
This command will place a new ResolvePage class within your app/GraphQL/Middleware directory.
```

mfn commented on src/Support/Field.php:
```diff
@@ -22,6 +24,9 @@ abstract class Field
 {
     protected $attributes = [];
 
+    /** @var string[] */
```
Please also add `@phpstan-var class-string[]`, thanks!

stevelacey replied:
I don't understand what this means, are you suggesting something like this?:

```suggestion
    /** @var class-string<Middleware> */
```

mfn replied:
Not replacing, but in addition ("also add"):
```php
@phpstan-var class-string[]
```
Actually I wouldn't add `Middleware` as a class, but technically we're doing duck-typing here => no one enforces that this has to be extending the abstract class we provide.

stevelacey replied:
I see. So like this? I see no examples in the project of any annotation like this

```suggestion
    /** 
     * @var string[]
     * @phpstan-var class-string[]
     */
```

mfn replied:
Exactly!

mfn commented on src/Support/Field.php:
```diff
@@ -94,7 +99,38 @@ public function getValidator(array $args, array $rules): ValidatorContract
+                    $result = $resolver($root, ...array_slice($arguments, 1));
+
+                    foreach ($this->middleware as $name) {
+                        $middleware = App::make($name);
```
Why are we using `app()` above and `App::make` here? (feels inconsistent)

I guess using the Facade would be preferred? (aka death to global functions)

stevelacey replied:
Yah I just copied this out of Laravel and never thought about it, facade it is.

mfn commented on src/Support/Middleware.php:
```diff
@@ -0,0 +1,25 @@
+        return $next($root, $args, $context, $info);
+    }
+
+    public function resolve(array $arguments, Closure $next)
```
Please add a phpdoc that this is called from `\Rebing\GraphQL\Support\Field::getResolver` (no one can statically analyze due to how the code has to work with the string name of this method :/)

stevelacey replied:
Something like this?

```suggestion
    /**
     * @see Field::getResolver()  Middleware is resolved in the field resolver pipeline
     */
    public function resolve(array $arguments, Closure $next)
```

stevelacey replied:
A related thing I considered was swapping the `handle` and `resolve` names around i.e. so the base class has `handle` that calls `resolve` instead of the current implementation `resolve` that calls `handle` – the default `via` pipelines use is `handle`.

I did it this way round so that end-user GraphQL middleware are more consistent with HTTP (by having a `handle` method) but I'm not sure if that's as useful as them having a `resolve` method to make it clear they intercept resolution rather than the HTTP stack.

mfn replied:
TBH that thought didn't cross my mind.

In fact it felt "natural" that it's the `handle` method one has to implement 🤷‍♀️

mfn replied:
Yeah, the `@see` seems appropriate!

stevelacey replied:
Handle makes sense to me too, I am happy to run with that

mfn commented on src/Support/Field.php:
```diff
@@ -94,7 +99,38 @@ public function getValidator(array $args, array $rules): ValidatorContract
     }
 
-    protected function getResolver(): ?Closure
+    protected function getResolver() : ?Closure
```
```suggestion
    protected function getResolver(): ?Closure
```

Guess that's a slap for php-cs-fixer, need to look into this

mfn commented on tests/Support/Objects/ExampleMiddleware.php:
```diff
@@ -0,0 +1,42 @@
+        }
+
+        if ($args['index'] === 5) {
+            throw new \Exception('Index 5 is not allowed');
```
Please always import, same also below

stevelacey commented:
@mfn addressed your comments and added some questions.

Yes, am interested in config/global definition, and maybe even better support for per-query middleware ala HTTP controller constructor `$this->middleware()` calls.

I didn't do that yet because I wondered if config was the right place for it, and because naming things is hard. HTTP middleware lives on the HTTP kernel, the best approximation of that might be adding it onto a `GraphQLServiceProvider`, which also mitigates the conflict with the existing HTTP middleware support in config somewhat. What do you think?

I figure the resolution order would be config, query class attr, constructor, and that'd be good enough for pretty much everyone. Laravel middleware does have the middleware ordering support as an additional feature they added later, we could add that if anyone ever wants it but I wouldn't say it's necessary for release, I've only ever really seen it used for auth anyway, which most will be doing at HTTP level.

mfn commented:
> but I wouldn't say it's necessary for release,

Ack 👍 and thanks for your insights!

mfn commented:
@stevelacey I think I answered all question, from my side it's ready to finalize this!

stevelacey commented:
@mfn what do you think about constructor level middleware additions (`$this->middleware(...)`) and global definition via service provider or similar instead of config? I feel like middleware refs belong more in app code than they do config, config is the one place Laravel doesn't put them.

mfn commented:
Wonder what's going on here 🤔
![image](https://user-images.githubusercontent.com/87493/76698916-35b7d380-66a8-11ea-83e0-d4a0e6ce8eef.png)

mfn commented:
> what do you think about constructor level middleware additions (`$this->middleware(...)`) and global definition via service provider or similar instead of config? I feel like middleware refs belong more in app code than they do config, config is the one place Laravel doesn't put them.

I think I would see some more specific proposal or sample code to "get a feel" for this.

middleware "config" is certainly not a Laravel pattern, but OTOH the GraphQL integration absolutely depends on the config.

I mean this could bet yet-another-things to consider: the GraphQL library suggests default paths/namespaces (`app/GraphQL/…`) so we would someone have to list all the types and mutations manually? Why wouldn't there be a mechanism which, like the console commands, can recursively load them all (if desired).

Certainly in my projects that would be welcome, my biggest has over 200 of such registered classes, all manually placed.

However: unless you want to make it the scope of this PR, I suggest to follow up on this rather and finalize this one.

mfn commented:
One use-case for this would be provide some kind of "filtering" based on permissions.

Ideally this would already be reflected in the schema but that's not always easily possible.

E.g. depending on user permissions, filtering out access to certain queries and mutations via this.

The `handle` method does not receive the query/mutation name itself but AFAIK it's available via `ResolveInfo` which should be good enough:
```php
class ExampleMiddleware extends Middleware
{
    public function handle($root, $args, $context, ResolveInfo $info, Closure $next)
    {
        var_dump($info->fieldName);
```
and the test output:
```
…
string(18) "examplesMiddleware"


Time: 290 ms, Memory: 14.00 MB

OK (1 test, 2 assertions)
```

mfn commented:
@stevelacey are you planning on moving forward?

Hope everything's fine due to current situation 🤞 

I'm also interested in the PR and I think everything is in place and I could even finish up the PR on my own but didn't want to just take it over.

stevelacey commented:
> I mean this could bet yet-another-things to consider: the GraphQL library suggests default paths/namespaces (`app/GraphQL/…`) so we would someone have to list all the types and mutations manually? Why wouldn't there be a mechanism which, like the console commands, can recursively load them all (if desired).

Yeah I'd like this too, again, a job for a service provider I think, we could copy the patterns established in `EventServiceProvider` aka `shouldDiscoverEvents` / `discoverEventsWithin`.

Another reason type discovery would be of benefit is by removing the problems created by the config pattern accepting array keys that must then be used when calling `GraphQL::type()`, instead of the type's actual name. This can create a real mess. I think a new type discovery shouldn't support this.

Something to think about later.

> One use-case for this would be provide some kind of "filtering" based on permissions.

Yeah kind of, though they'd still reflect in the introspection, it's not like they'd be truely filtered – authentication/authorization is one of my primary use-cases of middleware though.

> The `handle` method does not receive the query/mutation name itself but AFAIK it's available via `ResolveInfo` which should be good enough:

The `handle` method also receives the query/mutation instance as `$root`.

> @stevelacey are you planning on moving forward?
> 
> Hope everything's fine due to current situation 🤞

All is good, I plan to finish this off, but I can't commit to when, I'm on holiday this week, I'm interested in giving more thought to service provider / constructor middleware assignment as part of this work, I think it'd be good to keep the global definitions out of config, as a first step in potentially moving more things out and into a service provider discovery pattern.

mfn commented:
ping!

I released a new version now; wanted to wait for this, but not so long.

Still looking forward to this one!

mfn commented:
ping!

Are you still interested in moving forward?

I _might_ also be interesting wrapping this up (but haven't given this a deeper thought yet)

stevelacey commented:
I'll try to take a look within the next 30 days but if it doesn't happen by then you can safely assume this is up for grabs 👍

stevelacey commented:
@mfn okay this should be good to go, I didn't do anything more with global middlewares than make it slightly easier to override and customise the definition via a method, I think it's easiest to let users roll their own solution there – they can supply the `$middleware` attribute at any point in their inheritance hierarchy, or override `middleware()` and make it do something different.

radd2004 commented:
Ahh! excited for this to get merged in!

crissi commented:
still missing something here? and can I help?

mfn commented:
Time for review 😄 

I want to give this one a thorough check to ensure it's really useful, i.e. I've private project which could make use of this and simplify the current approach and I want to take the time to validate the approach works.

crissi commented:
> Time for review 😄
> 
> I want to give this one a thorough check to ensure it's really useful, i.e. I've private project which could make use of this and simplify the current approach and I want to take the time to validate the approach works.

okay, 

already found a couple of use cases that i need at work. 

- changing params in a query from int to string, without making the queries fail for the consumers of the api
- rate limiting specific mutations.

alanwillms commented:
This PR is awesome, I was manually intercepting `getResolver()` method so I could do my magic there.

rt3norio commented:
Hello!
Just for record:

Today we once again needed this feature.

When calling the Graphql endpoint with wrong parameters, the request ended even before the resolvers that was instrumented to grab user/IP/etc so we can identify/log the request and notify the third party client.

any forecast on this going live?

illambo commented:
Hi, any news about this PR ? Thanks

websitevirtuoso commented:
Yes. Do we have any news? Great feature

mfn reviewed (approved):
This is a truly stellar PR, absolutely great job @stevelacey 

Apologies for the super long delay, but 2020 wasn't an easy year so far…

Additionally, although the changes are not very intrusive, they go into the heart of the graphql code and I made it a requirement for my review to run this code on a private project which already was using a middleware-alike code and I wanted to know how easy/possible the transformation is => and it was super-easy, basically worked out of the box.

I found some super small issues I could not push in this PR due to missing permissions, so I just added them in https://github.com/rebing/graphql-laravel/pull/699

I did not touch your commits so I hope that merging #699 will also mark this one as merged 🤞 

Again, sorry for the long delay and thanks for the great work! 👍

mfn commented:
It worked 😅, PR is properly marked as merged.

mfn commented:
For anyone interested => you can test this via https://github.com/rebing/graphql-laravel/releases/tag/6.1.0-rc1

stevelacey commented:
@mfn great news 🎉 🙌
